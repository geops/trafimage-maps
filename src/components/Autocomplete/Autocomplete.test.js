import React from "react";
import { fireEvent, render } from "@testing-library/react";
// eslint-disable-next-line import/no-extraneous-dependencies
import userEvent from "@testing-library/user-event";
import Autocomplete from "./Autocomplete";

const defaultItems = [
  {
    label: "foo",
  },
  {
    label: "bar",
  },
  {
    label: "foo2",
  },
];

const items = [
  {
    label: "qux",
  },
  {
    label: "quux",
  },
  {
    label: "corge",
  },
];

let wrapper = null;

const mountDflt = () => {
  const { container } = render(
    <Autocomplete
      value="fooval"
      defaultItems={defaultItems}
      items={items}
      renderItem={(item) => item.label}
      getItemKey={(item) => item.label}
    />,
  );
  return container;
};

describe("Autocomplete", () => {
  describe("when no properties are set", () => {
    let spy = null;

    beforeEach(() => {
      window.console.error = jest.fn().mockImplementation(() => {});
      spy = jest.spyOn(window.console, "error");
    });

    afterEach(() => {
      spy.mockRestore();
      window.console.error.mockRestore();
      if (wrapper && wrapper.getDOMNode()) {
        document.body.removeChild(wrapper.getDOMNode());
        wrapper.unmount();
      }
      wrapper = null;
    });

    test("matches snapshot", () => {
      const { container } = render(<Autocomplete />);
      expect(container.innerHTML).toMatchSnapshot();
    });
  });

  describe("when properties are set", () => {
    test("should match snapshot without items and defaultItems using defaultProps", () => {
      const { container } = render(<Autocomplete value="fooval" />);
      container.querySelector("input").focus();
      expect(container.innerHTML).toMatchSnapshot();
    });

    test("should match snapshot without items and defaultItems", () => {
      const { container } = render(
        <Autocomplete
          value="fooval"
          renderTitle={() => "my_foo_title"}
          renderItem={(item) => item.label}
          getItemKey={() => Math.random()}
        />,
      );
      container.querySelector("input").focus();
      expect(container.innerHTML).toMatchSnapshot();
    });

    test("should match snapshot without defaultItems", () => {
      const { container } = render(
        <Autocomplete
          value="fooval"
          items={items}
          renderTitle={() => "my_foo_title"}
          renderItem={(item) => item.label}
          getItemKey={() => Math.random()}
        />,
      );
      container.querySelector("input").focus();
      expect(container.innerHTML).toMatchSnapshot();
    });

    test("should match snapshot without items", () => {
      const { container } = render(
        <Autocomplete
          value="fooval"
          defaultItems={defaultItems}
          renderTitle={() => "my_foo_title"}
          renderItem={(item) => item.label}
          getItemKey={() => Math.random()}
          onSelect={() => {}}
        />,
      );
      container.querySelector("input").focus();
      expect(container.innerHTML).toMatchSnapshot();
    });

    test("should match snapshot", () => {
      const { container } = render(
        <Autocomplete
          value="fooval"
          defaultItems={defaultItems}
          items={items}
          renderTitle={() => "my_foo_title"}
          renderItem={(item) => item.label}
          getItemKey={() => Math.random()}
          onSelect={() => {}}
        />,
      );
      expect(container.innerHTML).toMatchSnapshot();
    });

    test("should match snapshot showing list", () => {
      const { container } = render(
        <Autocomplete
          value="fooval"
          defaultItems={defaultItems}
          items={items}
          renderTitle={() => "my_foo_title"}
          renderItem={(item) => item.label}
          getItemKey={() => Math.random()}
          onSelect={() => {}}
        />,
      );
      container.querySelector("input").focus();
      expect(container.innerHTML).toMatchSnapshot();
    });

    describe("#onDocClick()", () => {
      test("hide the list and removes the focus.", () => {
        const container = mountDflt();
        container.querySelector("input").focus();
        fireEvent.click(document);
        expect(container.innerHTML).toMatchSnapshot();
      });

      test("does nothing when click comes from ana element of Autocomplete.", () => {
        const container = mountDflt();
        container.querySelector("input").focus();
        fireEvent.click(container.querySelector("div"));
        expect(container.innerHTML).toMatchSnapshot();
      });
    });

    describe("#onFocus()", () => {
      test("updates set showList state property to true", () => {
        const container = mountDflt();
        container.querySelector("input").focus();
        expect(container.querySelectorAll(".tm-focus").length).toBe(1);
      });
    });

    describe("#onBlurInput()", () => {
      test("does nothing if no lastKeyPress parameter", () => {
        const container = mountDflt();
        container.querySelector("input").focus();
        container.querySelector("input").blur();
        expect(container.querySelectorAll(".tm-focus").length).toBe(0);
      });

      test("hide list and renove focus if lastKeyPress is not an arrow down.", () => {
        const container = mountDflt();
        container.querySelector("input").focus();
        fireEvent.keyPress(container.querySelector("input"), { which: 41 });
        expect(container.querySelectorAll(".tm-focus").length).toBe(1);
      });
    });

    describe("#onKeyPress()", () => {
      test("gives focus to the first element of the list", () => {
        const container = mountDflt();
        container.querySelector("input").focus();
        fireEvent.keyUp(container.querySelector("input"), {
          keyCode: 40,
        });
        const li = container.querySelector("li");
        expect(li).toBe(document.activeElement);
      });

      test("doesn't give focus to the first element of the list", () => {
        const container = mountDflt();
        container.querySelector("input").focus();
        fireEvent.keyUp(container.querySelector("input"), {
          keyCode: 41,
        });
        const li = container.querySelector("li");
        expect(li).not.toBe(document.activeElement);
      });
    });

    describe("#onKeyPressItem()", () => {
      test("does nothing when another key than arrows is pressed.", () => {
        const container = mountDflt();
        const li = container.querySelectorAll("li")[1];
        li.focus();
        fireEvent.keyDown(li, {
          keyCode: 41,
        });
        expect(li).toBe(document.activeElement);
      });

      test("moves focus to search input.", () => {
        const container = mountDflt();
        const li = container.querySelectorAll("li")[0];
        li.focus();
        fireEvent.keyDown(li, {
          keyCode: 38,
        });
        expect(container.querySelector("input")).toBe(document.activeElement);
      });

      test("moves focus to previous item.", () => {
        const container = mountDflt();
        const lis = container.querySelectorAll("li");
        const li = lis[1];
        li.focus();
        fireEvent.keyDown(li, {
          keyCode: 38,
        });
        expect(lis[0]).toBe(document.activeElement);
      });

      test("moves focus to next item.", () => {
        const container = mountDflt();
        const lis = container.querySelectorAll("li");
        const li = lis[1];
        li.focus();
        fireEvent.keyDown(li, {
          keyCode: 40,
        });
        expect(lis[2]).toBe(document.activeElement);
      });

      test("moves focus to default items.", () => {
        const container = mountDflt();
        const lis = container.querySelectorAll("li");
        const li = lis[2];
        li.focus();
        fireEvent.keyDown(li, {
          keyCode: 40,
        });
        expect(lis[3]).toBe(document.activeElement);
      });

      test("moves focus to the beginning of the list.", () => {
        const container = mountDflt();
        const lis = container.querySelectorAll("li");
        const li = lis[5];
        li.focus();
        fireEvent.keyDown(li, {
          keyCode: 40,
        });
        expect(lis[0]).toBe(document.activeElement);
      });
    });

    describe("#onSelect()", () => {
      test("hide list and remove focus when select an item.", () => {
        const fn = jest.fn();
        const { container } = render(
          <Autocomplete
            value="fooval"
            defaultItems={defaultItems}
            items={items}
            renderItem={(item) => item.label}
            getItemKey={(item) => item.label}
            onSelect={fn}
          />,
        );
        container.querySelector("input").focus();
        fireEvent.click(container.querySelector("li"));
        expect(
          container.querySelector(".tm-autocomplete-results").style.display,
        ).toBe("none");
        expect(fn).toBeCalledTimes(1);
      });

      test("hide list and remove focus when select on default item.", () => {
        const fn = jest.fn();
        const { container } = render(
          <Autocomplete
            value="fooval"
            defaultItems={defaultItems}
            items={items}
            renderItem={(item) => item.label}
            getItemKey={(item) => item.label}
            onSelect={fn}
          />,
        );
        container.querySelector("input").focus();
        fireEvent.click(container.querySelectorAll("li")[4]);
        expect(
          container.querySelector(".tm-autocomplete-results").style.display,
        ).toBe("none");
        expect(fn).toBeCalledTimes(1);
      });
    });

    describe("#onChange()", () => {
      test("gives focus to the input and display the list.", async () => {
        const user = userEvent.setup();
        const fn = jest.fn();
        const { container } = render(
          <Autocomplete
            value="fooval"
            defaultItems={defaultItems}
            items={items}
            renderItem={(item) => item.label}
            getItemKey={(item) => item.label}
            onChange={fn}
          />,
        );
        container.querySelector("input").focus();
        await user.keyboard("foo");
        expect(fn).toBeCalledTimes(3);
      });
    });

    test("hides the list on click on search button if the list is open.", () => {
      const container = mountDflt();
      container.querySelector("input").focus();
      expect(
        container.querySelector(".tm-autocomplete-results").style.display,
      ).toBe("");
      fireEvent.click(container.querySelectorAll('[role="button"]')[1]);
      expect(
        container.querySelector(".tm-autocomplete-results").style.display,
      ).toBe("none");
    });
  });
});
