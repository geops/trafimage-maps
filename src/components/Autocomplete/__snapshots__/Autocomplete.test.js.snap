// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete when no properties are set matches snapshot 1`] = `
<div class="tm-autocomplete">
  <div class="tm-search-input">
    <input type="text"
           tabindex="0"
           placeholder
           title
           value
    >
    <div role="button"
         style="display: none;"
         tabindex="0"
         class="tm-button tm-bt-clear"
         title
         aria-label
    >
      <svg stroke="currentColor"
           fill="currentColor"
           stroke-width="0"
           viewbox="0 0 24 24"
           focusable="false"
           height="1em"
           width="1em"
           xmlns="http://www.w3.org/2000/svg"
      >
        <path fill="none"
              d="M0 0h24v24H0z"
        >
        </path>
        <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z">
        </path>
      </svg>
    </div>
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-search"
         title
         aria-label
    >
      submit
    </div>
  </div>
  <div style="display: none;"
       class="tm-autocomplete-results"
  >
  </div>
</div>
`;

exports[`Autocomplete when properties are set #onDocClick() does nothing when click comes from ana element of Autocomplete. 1`] = `
<div class="tm-autocomplete">
  <div class="tm-search-input tm-focus">
    <input type="text"
           tabindex="0"
           placeholder
           title
           value="fooval"
    >
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-clear"
         title
         aria-label
    >
      <svg stroke="currentColor"
           fill="currentColor"
           stroke-width="0"
           viewbox="0 0 24 24"
           focusable="false"
           height="1em"
           width="1em"
           xmlns="http://www.w3.org/2000/svg"
      >
        <path fill="none"
              d="M0 0h24v24H0z"
        >
        </path>
        <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z">
        </path>
      </svg>
    </div>
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-search"
         title
         aria-label
    >
      submit
    </div>
  </div>
  <div style
       class="tm-autocomplete-results"
  >
    <ul role="menu"
        class="tm-list"
    >
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        qux
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        quux
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        corge
      </li>
    </ul>
    <hr>
    <ul role="menu"
        class="tm-list"
    >
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        foo
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        bar
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        foo2
      </li>
    </ul>
  </div>
</div>
`;

exports[`Autocomplete when properties are set #onDocClick() hide the list and removes the focus. 1`] = `
<div class="tm-autocomplete">
  <div class="tm-search-input tm-focus">
    <input type="text"
           tabindex="0"
           placeholder
           title
           value="fooval"
    >
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-clear"
         title
         aria-label
    >
      <svg stroke="currentColor"
           fill="currentColor"
           stroke-width="0"
           viewbox="0 0 24 24"
           focusable="false"
           height="1em"
           width="1em"
           xmlns="http://www.w3.org/2000/svg"
      >
        <path fill="none"
              d="M0 0h24v24H0z"
        >
        </path>
        <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z">
        </path>
      </svg>
    </div>
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-search"
         title
         aria-label
    >
      submit
    </div>
  </div>
  <div style="display: none;"
       class="tm-autocomplete-results"
  >
    <ul role="menu"
        class="tm-list"
    >
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        qux
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        quux
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        corge
      </li>
    </ul>
    <hr>
    <ul role="menu"
        class="tm-list"
    >
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        foo
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        bar
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        foo2
      </li>
    </ul>
  </div>
</div>
`;

exports[`Autocomplete when properties are set should match snapshot 1`] = `
<div class="tm-autocomplete">
  <div class="tm-search-input">
    <input type="text"
           tabindex="0"
           placeholder
           title
           value="fooval"
    >
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-clear"
         title
         aria-label
    >
      <svg stroke="currentColor"
           fill="currentColor"
           stroke-width="0"
           viewbox="0 0 24 24"
           focusable="false"
           height="1em"
           width="1em"
           xmlns="http://www.w3.org/2000/svg"
      >
        <path fill="none"
              d="M0 0h24v24H0z"
        >
        </path>
        <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z">
        </path>
      </svg>
    </div>
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-search"
         title
         aria-label
    >
      submit
    </div>
  </div>
  <div style="display: none;"
       class="tm-autocomplete-results"
  >
    <ul role="menu"
        class="tm-list"
    >
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        qux
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        quux
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        corge
      </li>
    </ul>
    <hr>
    my_foo_title
    <ul role="menu"
        class="tm-list"
    >
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        foo
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        bar
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        foo2
      </li>
    </ul>
  </div>
</div>
`;

exports[`Autocomplete when properties are set should match snapshot showing list 1`] = `
<div class="tm-autocomplete">
  <div class="tm-search-input tm-focus">
    <input type="text"
           tabindex="0"
           placeholder
           title
           value="fooval"
    >
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-clear"
         title
         aria-label
    >
      <svg stroke="currentColor"
           fill="currentColor"
           stroke-width="0"
           viewbox="0 0 24 24"
           focusable="false"
           height="1em"
           width="1em"
           xmlns="http://www.w3.org/2000/svg"
      >
        <path fill="none"
              d="M0 0h24v24H0z"
        >
        </path>
        <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z">
        </path>
      </svg>
    </div>
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-search"
         title
         aria-label
    >
      submit
    </div>
  </div>
  <div style
       class="tm-autocomplete-results"
  >
    <ul role="menu"
        class="tm-list"
    >
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        qux
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        quux
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        corge
      </li>
    </ul>
    <hr>
    my_foo_title
    <ul role="menu"
        class="tm-list"
    >
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        foo
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        bar
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        foo2
      </li>
    </ul>
  </div>
</div>
`;

exports[`Autocomplete when properties are set should match snapshot without defaultItems 1`] = `
<div class="tm-autocomplete">
  <div class="tm-search-input tm-focus">
    <input type="text"
           tabindex="0"
           placeholder
           title
           value="fooval"
    >
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-clear"
         title
         aria-label
    >
      <svg stroke="currentColor"
           fill="currentColor"
           stroke-width="0"
           viewbox="0 0 24 24"
           focusable="false"
           height="1em"
           width="1em"
           xmlns="http://www.w3.org/2000/svg"
      >
        <path fill="none"
              d="M0 0h24v24H0z"
        >
        </path>
        <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z">
        </path>
      </svg>
    </div>
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-search"
         title
         aria-label
    >
      submit
    </div>
  </div>
  <div style
       class="tm-autocomplete-results"
  >
    <ul role="menu"
        class="tm-list"
    >
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        qux
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        quux
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        corge
      </li>
    </ul>
    my_foo_title
  </div>
</div>
`;

exports[`Autocomplete when properties are set should match snapshot without items 1`] = `
<div class="tm-autocomplete">
  <div class="tm-search-input tm-focus">
    <input type="text"
           tabindex="0"
           placeholder
           title
           value="fooval"
    >
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-clear"
         title
         aria-label
    >
      <svg stroke="currentColor"
           fill="currentColor"
           stroke-width="0"
           viewbox="0 0 24 24"
           focusable="false"
           height="1em"
           width="1em"
           xmlns="http://www.w3.org/2000/svg"
      >
        <path fill="none"
              d="M0 0h24v24H0z"
        >
        </path>
        <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z">
        </path>
      </svg>
    </div>
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-search"
         title
         aria-label
    >
      submit
    </div>
  </div>
  <div style
       class="tm-autocomplete-results"
  >
    my_foo_title
    <ul role="menu"
        class="tm-list"
    >
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        foo
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        bar
      </li>
      <li class="tm-list-item"
          role="menuitem"
          tabindex="0"
      >
        foo2
      </li>
    </ul>
  </div>
</div>
`;

exports[`Autocomplete when properties are set should match snapshot without items and defaultItems 1`] = `
<div class="tm-autocomplete">
  <div class="tm-search-input tm-focus">
    <input type="text"
           tabindex="0"
           placeholder
           title
           value="fooval"
    >
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-clear"
         title
         aria-label
    >
      <svg stroke="currentColor"
           fill="currentColor"
           stroke-width="0"
           viewbox="0 0 24 24"
           focusable="false"
           height="1em"
           width="1em"
           xmlns="http://www.w3.org/2000/svg"
      >
        <path fill="none"
              d="M0 0h24v24H0z"
        >
        </path>
        <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z">
        </path>
      </svg>
    </div>
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-search"
         title
         aria-label
    >
      submit
    </div>
  </div>
  <div style="display: none;"
       class="tm-autocomplete-results"
  >
    my_foo_title
  </div>
</div>
`;

exports[`Autocomplete when properties are set should match snapshot without items and defaultItems using defaultProps 1`] = `
<div class="tm-autocomplete">
  <div class="tm-search-input tm-focus">
    <input type="text"
           tabindex="0"
           placeholder
           title
           value="fooval"
    >
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-clear"
         title
         aria-label
    >
      <svg stroke="currentColor"
           fill="currentColor"
           stroke-width="0"
           viewbox="0 0 24 24"
           focusable="false"
           height="1em"
           width="1em"
           xmlns="http://www.w3.org/2000/svg"
      >
        <path fill="none"
              d="M0 0h24v24H0z"
        >
        </path>
        <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z">
        </path>
      </svg>
    </div>
    <div role="button"
         tabindex="0"
         class="tm-button tm-bt-search"
         title
         aria-label
    >
      submit
    </div>
  </div>
  <div style="display: none;"
       class="tm-autocomplete-results"
  >
  </div>
</div>
`;
